File: src/pokemon.cpp
#include "pokemon.hpp"

#include <iostream>
#include <string>

// Public Methods

Pokemon::Pokemon(const std::string &name, float attack, float defense, float hp,
                 const std::string &type)
    : m_name(std::move(name)), m_attack(attack), m_defense(defense), m_HP(hp),
      m_type(std::move(type)), m_superEffective(defineSuperEffective()) {}

void Pokemon::Attack(Pokemon &other) const {
  float multiplier = getAttackMultiplier(other);
  float dmg = m_attack * multiplier - other.m_defense;

  if (dmg <= 0)
    dmg = 1;

  other.m_HP -= dmg;
}

std::string Pokemon::getName() const { return m_name; }
float Pokemon::getAttack() const { return m_attack; }
float Pokemon::getDefense() const { return m_defense; }
float Pokemon::getHP() const { return m_HP; }
std::string Pokemon::getType() const { return m_type; }
std::string Pokemon::getSuperEffective() const { return m_superEffective; }

// Private Methods

std::string Pokemon::defineSuperEffective(void) const {
  std::string types[] = {"electric", "water", "fire", "ice", "rock"};
  std::string counterTypes[] = {"water", "fire", "ice", "rock", "electric"};

  for (int i = 0; i < 5; i++) {
    if (m_type == types[i])
      return counterTypes[i];
  }
  std::cerr << "Type '" << m_type << "' not found, leaving program."
            << std::endl;
  exit(1);
}

float Pokemon::getAttackMultiplier(const Pokemon &other) const {
  float multiplier = 1;

  if (this->m_superEffective == other.m_type)
    multiplier = 1.2;

  if (other.m_superEffective == this->m_type)
    multiplier = 0.8;

  return multiplier;
}

File: src/pokemon.rs
use crate::utils::{PokemonResult, get_super_effective};

#[derive(Debug)]
pub struct Pokemon {
    name: String,
    typ: String,
    super_effective: String,
    attack: f32,
    defense: f32,
    hp: f32,
}

/// Public methods
impl Pokemon {
    pub fn new(
        name: String,
        typ: String,
        attack: f32,
        defense: f32,
        hp: f32,
    ) -> PokemonResult<Self> {
        let super_effective = get_super_effective(&typ)?;
        Ok(Self {
            name,
            typ,
            super_effective,
            attack,
            defense,
            hp,
        })
    }

    pub fn attack(&self, defense: &mut Pokemon) {
        let multiplier = self.get_attack_multiplier(defense);
        let dmg = {
            let tmp = self.attack * multiplier - defense.defense;
            if tmp <= 0.0 { 1.0 } else { tmp }
        };

        defense.hp -= dmg;
    }

    pub fn hp(&self) -> f32 {
        self.hp
    }

    pub fn alive(&self) -> bool {
        self.hp > 0.0
    }

    pub fn name(&self) -> &str {
        &self.name
    }
}

/// Private methods
impl Pokemon {
    fn get_attack_multiplier(&self, defense: &Pokemon) -> f32 {
        if self.super_effective == defense.typ {
            1.2
        } else if defense.super_effective == self.typ {
            0.8
        } else {
            1.0
        }
    }
}
